### PR TITLE
Robust Dpatch

### PR DESCRIPTION
**Pull request by mathsinn**
# Description

Implementation of new class towards generating robust physical-world adversarial patches for object detectors.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Local proofs of concept (Jupyter notebook and example script)
- [x] Unit tests

**Test Configuration**:
- MacOS
- Python 3.7
- ART version 1.4.2
- PyTorch 1.6.0

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/dpatch_robust.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+        else:
+            image_height, image_width = x.shape[1:3]
+
+        if (
+            self.patch_location[0] + self.patch_shape[0] > image_height - self.crop_range[0]
+            or self.patch_location[1] + self.patch_shape[1] > image_width - self.crop_range[1]
+        ):
+            raise ValueError("The patch (partially) lies outside the cropped image.")
+
+        for i_step in trange(self.max_iter, desc="DPatchEOT iteration", disable=not self.verbose):
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:154]:
```suggestion
        for i_step in trange(self.max_iter, desc="RobustDPatch iteration", disable=not self.verbose):
```

File: art/attacks/evasion/dpatch_robust.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+                    a_min=self.estimator.clip_values[0],
+                    a_max=self.estimator.clip_values[1],
+                )
+
+        return self._patch
+
+    @deprecated_keyword_arg("channel_index", end_version="1.5.0", replaced_by="channels_first")
+    def _augment_images_with_patch(
+        self, x: np.ndarray, patch: np.ndarray, channels_first: bool, channel_index=Deprecated
+    ) -> Tuple[np.ndarray, np.ndarray, Dict[str, Union[int, float]]]:
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:205]:
```suggestion
    def _augment_images_with_patch(
        self, x: np.ndarray, patch: np.ndarray, channels_first: bool
    ) -> Tuple[np.ndarray, np.ndarray, Dict[str, Union[int, float]]]:
```

File: art/attacks/evasion/dpatch_robust.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        """
+        Augment images with patch.
+
+        :param x: Sample images.
+        :param patch: The patch to be applied.
+        :param channels_first: Set channels first or last.
+        :param channel_index: Index of the color channel.
+        :type channel_index: `int`
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:213]:
```suggestion
```

File: art/attacks/evasion/dpatch_robust.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        """
+        # Remove in 1.5.0
+        if channel_index == 3:
+            channels_first = False
+        elif channel_index == 1:
+            channels_first = True
+        elif channel_index is not Deprecated:
+            raise ValueError("Not a proper channel_index. Use channels_first.")
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:221]:
```suggestion
```

File: art/attacks/evasion/dpatch_robust.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+        return x_patch, patch_target, transformations
+
+    @deprecated_keyword_arg("channel_index", end_version="1.5.0", replaced_by="channels_first")
+    def _untransform_gradients(
+        self,
+        gradients: np.ndarray,
+        transforms: Dict[str, Union[int, float]],
+        channels_first: bool,
+        channel_index=Deprecated,
+    ) -> np.ndarray:
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:287]:
```suggestion
    def _untransform_gradients(
        self,
        gradients: np.ndarray,
        transforms: Dict[str, Union[int, float]],
        channels_first: bool,
    ) -> np.ndarray:
```

File: art/attacks/evasion/dpatch_robust.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+        """
+        Augment images with patch.
+
+        :param x: Sample images.
+        :param patch: The patch to be applied.
+        :param channels_first: Set channels first or last.
+        :param channel_index: Index of the color channel.
+        :type channel_index: `int`
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:295]:
```suggestion
```

File: art/attacks/evasion/dpatch_robust.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+        """
+        # Remove in 1.5.0
+        if channel_index == 3:
+            channels_first = False
+        elif channel_index == 1:
+            channels_first = True
+        elif channel_index is not Deprecated:
+            raise ValueError("Not a proper channel_index. Use channels_first.")
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:303]:
```suggestion
```

File: tests/attacks/evasion/test_robust_dpatch.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+import numpy as np
+import pytest
+
+from art.attacks.evasion import RobustDPatch
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.object_detection.object_detector import ObjectDetectorMixin, ObjectDetector
+from tests.attacks.utils import backend_test_classifier_type_check_fail
+from tests.utils import master_seed
+from tests.utils import ARTTestException
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_robust_dpatch.py:28]:
```suggestion
from art.attacks.evasion import RobustDPatch
from art.estimators.estimator import BaseEstimator, LossGradientsMixin
from art.estimators.object_detection.object_detector import ObjectDetectorMixin, ObjectDetector

from tests.attacks.utils import backend_test_classifier_type_check_fail
from tests.utils import master_seed
from tests.utils import ARTTestException
```

File: art/attacks/evasion/dpatch_robust.py [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+import random
+from typing import Dict, List, Optional, Tuple, Union, TYPE_CHECKING
+
+import numpy as np
+from tqdm import trange
+
+from art.attacks.attack import EvasionAttack
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.object_detection.object_detector import ObjectDetectorMixin
+from art.utils import Deprecated, deprecated_keyword_arg
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:38]:
```suggestion
```

File: art/attacks/evasion/dpatch_robust.py [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+                * (self.estimator.clip_values[1] - self.estimator.clip_values[0])
+                + self.estimator.clip_values[0]
+            ).astype(config.ART_NUMPY_DTYPE)
+        self.verbose = verbose
+        self.patch_location = patch_location
+        self.crop_range = crop_range
+        self.brightness_range = brightness_range
+        self.rotation_weights = rotation_weights
+        self.sample_size = sample_size
+        super()._check_params()
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:123]:
```suggestion
```

File: art/attacks/evasion/dpatch_robust.py [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+
+        :param x: Sample images.
+        :param y: Target labels for object detector.
+        :return: Adversarial patch.
+        """
+        channel_index = 1 if self.estimator.channels_first else x.ndim - 1
+        if x.shape[channel_index] != self.patch_shape[channel_index - 1]:
+            raise ValueError("The color channel index of the images and the patch have to be identical.")
+        if y is not None:
+            raise ValueError("The DPatchEOT attack does not use target labels.")
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:138]:
```suggestion
            raise ValueError("The RobustDPatch attack does not use target labels.")
```

File: art/attacks/evasion/dpatch_robust.py [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+        if x.ndim != 4:
+            raise ValueError("The adversarial patch can only be applied to images.")
+
+        # Check whether patch fits into the cropped images:
+        if channel_index == 1:
+            image_height, image_width = x.shape[2:4]
+        else:
+            image_height, image_width = x.shape[1:3]
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:146]:
```suggestion
        if self.estimator.channels_first:
            image_height, image_width = x.shape[2:4]
        else:
            image_height, image_width = x.shape[1:3]
```

File: art/attacks/evasion/dpatch_robust.py (hunk 2 above)
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:205]:
```suggestion
    ) -> Tuple[np.ndarray, List[Dict[str, np.ndarray]], Dict[str, Union[int, float]]]:
```

File: art/attacks/evasion/dpatch_robust.py [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+        gradients: np.ndarray,
+        transforms: Dict[str, Union[int, float]],
+        channels_first: bool,
+        channel_index=Deprecated,
+    ) -> np.ndarray:
+        """
+        Augment images with patch.
+
+        :param x: Sample images.
+        :param patch: The patch to be applied.
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:292]:
```suggestion
        Revert transformation on gradients.

        :param gradients: The gradients to be reverse transformed.
        :param transforms: The transformations in forward direction.
```

File: tests/attacks/evasion/test_robust_dpatch.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
+# MIT License
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_robust_dpatch.py:1]: Should we rename this file to `test_dpatch_robust.py` following the convention of prepending `test_` to the tested module's filename?

File: tests/attacks/evasion/test_robust_dpatch.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import logging
+
+import numpy as np
+import pytest
+
+from art.attacks.evasion import RobustDPatch
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.object_detection.object_detector import ObjectDetectorMixin, ObjectDetector
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_robust_dpatch.py:25]:
```suggestion
from art.estimators.object_detection.object_detector import ObjectDetectorMixin
```

File: tests/attacks/evasion/test_robust_dpatch.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+from tests.attacks.utils import backend_test_classifier_type_check_fail
+from tests.utils import master_seed
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_robust_dpatch.py:27]:
```suggestion
```

File: art/attacks/evasion/dpatch_robust.py [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+    _estimator_requirements = (BaseEstimator, LossGradientsMixin, ObjectDetectorMixin)
+
+    def __init__(
+        self,
+        estimator: "OBJECT_DETECTOR_TYPE",
+        patch_shape: Tuple[int, int, int] = (40, 40, 3),
+        patch_location: Tuple[int, int] = (0, 0),
+        crop_range: Tuple[int, int] = (0, 0),
+        brightness_range: Tuple[float, float] = (1.0, 1.0),
+        rotation_weights: Union[Tuple[float], Tuple[int]] = (1, 0, 0, 0),
```
Comment (beat-buesser) [inline art/attacks/evasion/dpatch_robust.py:79]:
```suggestion
        rotation_weights: Union[Tuple[float, float, float, float], Tuple[int, int, int, int]] = (1, 0, 0, 0),
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @mathsinn Thank you very much for implementing and testing a better, more robust version of  `DPatch`! I think it's a great name for the new attack. I have proposed a few updates above.

File: tests/attacks/evasion/test_robust_dpatch.py (hunk 14 above)
Comment (mathsinn) [reply tests/attacks/evasion/test_robust_dpatch.py]: I agree; I've renamed the file.

**Comment by mathsinn** [comment]
Thank you @beat-buesser for the thorough review and the suggested changes. I've incorporated them all and also renamed the test module as per your suggestion.

**Review by beat-buesser — APPROVED** [review]

**Pull request merged**